README.md
# Tds.Ecto (Ecto v2)

MSSQL / TDS Adapter for Ecto2

[![Build Status][appveyor-img]][appveyor] [![Hex Version][hex-img]][hex]

[appveyor-img]: https://ci.appveyor.com/api/projects/status/g59ocaellinuig7g?svg=true
[appveyor]: https://ci.appveyor.com/project/mjaric/tds-ecto-jpd3h
[hex-img]: https://img.shields.io/hexpm/v/tds_ecto.svg
[hex]: https://hex.pm/packages/tds_ecto

## Example
```elixir
# In your config/config.exs file add MSSQL connection options
# By default it use port 1433, different port can be specified with
# port: 1434
# option and named instance with
# instance: "instance_name"
config :my_app, Repo,
  database: "ecto_simple",
  username: "mssql",
  password: "mssql",
  hostname: "localhost"

# In your application code
defmodule Repo do
  use Ecto.Repo,
    otp_app: :my_app,
    adapter: Tds.Ecto
end

defmodule Weather do
  use Ecto.Schema

  schema "weather" do
    field :city     # Defaults to type :string
    field :temp_lo, :integer
    field :temp_hi, :integer
    field :prcp,    :float, default: 0.0
  end
end

defmodule Simple do
  import Ecto.Query

  def sample_query do
    query = from w in Weather,
          where: w.prcp > 0 or is_nil(w.prcp),
         select: w
    Repo.all(query)
  end
end
```

Also, if you need to map table to schema other than [dbo], simple use @schema_prefix "your_schema", eg:

```
defmodule Invoices.Invoice do
  use Ecto.Schema

  @schema_prefix :invoices
  schema "invoices" do
    field :due_date, :datetime
    field :sum, :float
  end
end
```

## Usage

Add Tds as a dependency in your `mix.exs` file.

```elixir
def deps do
  [{:tds_ecto, "~> 2.2"}]
end
```

You should also update your applications list to include both projects:
```elixir
def application do
  [applications: [:logger, :tds_ecto, :ecto]]
end
```

To use the adapter in your repo:
```elixir
defmodule MyApp.Repo do
  use Ecto.Repo,
    otp_app: :my_app,
    adapter: Tds.Ecto
end
```

tds_ecto relies on pattern matching against error messages to extract constraint names.
If your SQL server is not configured to English as its default language you may add an `after_connect` hook to your repo and set English as the language for Ecto connections.

```elixir
def after_connect(conn) do
  Tds.Ecto.Connection.query(conn, "SET LANGUAGE English", [], [])
end
```

For additional information on usage please see the documentation for [Ecto](http://hexdocs.pm/ecto).

## Data Type Mapping

    MSSQL             	Ecto/Tds.Ecto
    ----------        	------
    nvarchar            :string
    varchar             Tds.VarChar
    char                :binary
    varbinary           :binary
    float               :float
    decimal             :decimal
    integer             :integer
    bit                 :boolean
    uniqueidentifier    :uuid
    datetime            :datetime
    date                :date
    time                :time



## Contributing

To contribute you need to compile Tds from source and test it:

```
$ git clone https://github.com/livehelpnow/tds_ecto.git
$ cd tds_ecto
$ mix test
```

Tests will try to connect to `localhost` using `sa` as username with predefined password `mssql`, but you can set environment variables to override any, like so: `SQL_USERNAME=myuser; SQL_PASSWORD=mypassword; SQL_HOSTNAME=sqlserver.local; mix test` or if you are using windows then `SET SQL_USERNAME=myuser && SET SQL_PASSWORD=mypassword && SET SQL_HOSTNAME=sqlserver.local && mix test`. Please note that tests will run againes default sql server instance.

Additionally SQL authentication needs to be used for connecting and testing.If you override default test settings, make sure either that yor user has sysadmin privilegies to add the user `test_user` with access to the database `test_db`, or simply add manualy test_user on your server. See one of the test files for the connection information and port number.

## License

   Copyright 2014, 2015, 2016, 2017 LiveHelpNow

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.


CHANGELOG.md
#v2.2.1
* Fixing transaction handling (ecto cast_assoc rollsback twice if more than one rpc call is executed in single transaction/change/assoc)

#v2.2.0
## Breaking Changes
UUID/UNIQUEIDENTIFER column is now stored AS IS in database, meaning that compatibility with Ecto.UUID is broken, 
if you are using `tds_ecto` please use Tds.UUID to encode/decode value to its sting representation. MSSQL has its way of 
parsing binary UUID value into string representation as following string illustration:

  Ecto.UUID string representation:
  `f72f71ce-ee18-4db3-74d9-f5662a7691b8`

  MSSQL string representation:
  `ce712ff7-18ee-b34d-74d9-f5662a7691b8`

To allow other platforms to interprect corectly uuids we had to introduce 
`Tds.UUID` in `tds_ecto` library and `Tds.Types.UUID` in `tds` both are trying to 
keep binary storage in valid byte order so each platform can corectly decode it into string.
So far unique identifiers were and will be returned in resultset as binary, if you need to convert it into 
formatted string, use `Tds.Types.UUID.parse(<<_::128>>=uuid)` to get that string. 
It is safe to call this function several times since it will not fail if value is valid uuid string,
it will just return same value. But if value do not match `<<_::128>>` or 
`<<a::32, ?-, b::16, ?-, c::16, ?- d::16, ?-, e::48>>` it will throw runtime error. 

If you are using `tds_ecto` :uuid, :binary_id, and Tds.UUID are types you want to use in your models.
For any of those 3 types auto encode/decode will be performed.

Since there was a bug where in some cases old version of `tds_ecto` library 
could not detemine if binary is of uuid type, it interpeted such values as raw binary which caused some issues when non elixir apps interpreted that binary in wrong string format. This was ok as long as parsed string values were not shared between elixir and other platforms trough e.g. json messages, but if they did, this could be a problem where other app is not kapable to find object which missparsed uuid value.

#v2.1.0
* Introducing Tds.VarChar type, use in migration `:varchar` as field type and in schema `Tds.VarChar` as field type.

# v2.0.7
* fixing issue when inserting data with decimal columns see issue #57

# v2.0.3 (v2.0.2)
* Fixing failed test when schemaless query is executed without specified columns in projection

# v1.0.6 (v1.0.2)
* Integrated upstream changes
* Add default expression to alter tables too.

# v1.0.4 (v1.0.2)
* Integrated upstream changes

# v1.0.3
* Bug fixes
  * Rollbacks for columns with defaults is now working.
  * Forwards for columns with defaults is now working.

# v1.0.2
* Bug Fixes
  * fixed limit

# v1.0.1
* Bug Fixes
  * Allow prefixes in DDL
  * Fixed issue when using `count` and `distinct`
  * Better constraint matching
  * Fixed boolean literals in where clauses

# v1.0.0
* Enhancements
  * Updated to ecto 1.0.0

# v0.6.0
* Enhancements
  * Updated to ecto 0.16.0

# v0.5.0
* Enhancements
  * Updated to ecto 0.15.0

# v0.4.0
* Enhancements
  * Updated to ecto 0.14.1

# v0.3.2
* Enhancements
  * Lock down tds

# v0.3.1
* Enhancements
  * Updated tds to 0.4.0

# v0.3.0
* Enhancements
  * Updated to ecto 0.13.0


# v0.2.5
* Enhancements
  * Updated Ecto to 0.12.1
  * Added support for map type
  * Added support for binary_id

# v0.2.4
* Enhancements
  * Updated Ecto to 0.11

# v0.2.3
* Bug Fixes
  * Changed tablename quoting to allow for schema

# v0.2.2
* Enhancements
  * Updated Ecto version

# v0.2.1
* Bug Fixes
  * Use datetime2 when time struct usec > 0

# v0.2.0
* Bug Fixes
  * Parsing null uuid's would break String check
  * Allow references to be set for other primary key types
  * Allow distincts on booleans

# v0.1.5
* Bug Fixes
  * Ecto.UUID tagged types were failing to be type checked in tds

* Enhancements
  * Added support for UNIQUE in column definitions

# v0.1.4
* Bug Fixes
  * Added :tds to application list

# v0.1.3
* Bug Fixes
  * Fixing deps issue with Hex

# v0.1.2
* Bug Fixes
  * Added lock to join/2

# v0.1.1
* Bug Fixes
  * Fixed limit: to map TOP in SELECT

# v0.1.0 (2015-02-13)
* Enhancements
  * Updated Deps


# v0.0.1 (2015-02-12)
* First Release


config/test.exs
use Mix.Config

config :logger, level: :info

config :ecto, lock_for_update: true

config :tds_ecto,
  opts: [
    hostname: "localhost",
    username: "sa",
    password: System.get_env("SQL_PASSWORD") || "some!Password",
    database: "test",
    set_allow_snapshot_isolation: :on
  ]



test/test_helper.exs
Logger.configure(level: :info)

ExUnit.start(
  exclude: [
    :assigns_id_type,
    :array_type,
    :case_sensitive,
    :modify_foreign_key_on_update,
    :modify_foreign_key_on_delete,
    :uses_usec,
    :lock_for_update,
    # NOTE: there is bug with transaction timout, it works bt error is not returned Repo transaction function
    :disconnect_on_transaction_timout,
    # Note: this requires read uncommited, otherwise it is normal to lock :)
    :transactions_are_not_shared_in_repo,

    :with_conflict_target, :with_conflict_ignore
  ]
)

Application.put_env(:ecto, :lock_for_update, "FOR UPDATE")
Application.put_env(:ecto, :primary_key_type, :id)
Application.put_env(:ecto, :async_integration_tests, false)

# Basic test repo
Code.require_file("./integration/support/types.exs", __DIR__)
Code.require_file("./integration/support/repo.exs", __DIR__)
Code.require_file("./integration/support/schemas.exs", __DIR__)
Code.require_file("./integration/support/migration.exs", __DIR__)

pool =
  case System.get_env("ECTO_POOL") || "poolboy" do
    "poolboy" -> DBConnection.Poolboy
    "sbroker" -> DBConnection.Sojourn
  end

alias Ecto.Integration.TestRepo

Application.put_env(
  :ecto,
  TestRepo,
  filter_null_on_unique_indexes: true,
  adapter: Tds.Ecto,
  hostname: System.get_env("SQL_HOSTNAME") || "localhost",
  username: System.get_env("SQL_USERNAME") || "sa",
  password: System.get_env("SQL_PASSWORD") || "some!Password",
  database: "ecto_test",
  pool: Ecto.Adapters.SQL.Sandbox,
  ownership_pool: pool,
  set_allow_snapshot_isolation: :on
)

defmodule Ecto.Integration.TestRepo do
  use Ecto.Integration.Repo, otp_app: :ecto
end

alias Ecto.Integration.PoolRepo

Application.put_env(
  :ecto,
  PoolRepo,
  adapter: Tds.Ecto,
  pool: pool,
  hostname: System.get_env("SQL_HOSTNAME") || "localhost",
  username: System.get_env("SQL_USERNAME") || "sa",
  password: System.get_env("SQL_PASSWORD") || "some!Password",
  database: "ecto_test",
  max_restarts: 20,
  max_seconds: 10,
  pool_size: 10,
  set_allow_snapshot_isolation: :on
)

defmodule Ecto.Integration.PoolRepo do
  use Ecto.Integration.Repo, otp_app: :ecto

  def create_prefix(prefix) do
    "create schema #{prefix}"
  end

  def drop_prefix(prefix) do
    "drop schema #{prefix}"
  end
end

defmodule Ecto.Integration.Case do
  use ExUnit.CaseTemplate

  alias Ecto.Integration.TestRepo

  setup do
    :ok = Ecto.Adapters.SQL.Sandbox.checkout(TestRepo)
  end
end

# :debugger.start()
# :int.ni(Tds.Ecto)
# :int.break(Tds.Ecto, 164)

:dbg.tracer()
:dbg.p(:all, :c)
# :dbg.tpl(Tds.Protocol, :handle_begin, :x)
# :dbg.tpl(Tds.Protocol, :handle_commit, :x)
# :dbg.tpl(Tds.Protocol, :handle_rollback, :x)
# :dbg.tpl(Tds.Protocol, :handle_prepare, :x)
# :dbg.tpl(Tds.Protocol, :handle_exec, :x)

:erlang.system_flag(:backtrace_depth, 50)

{:ok, _} = Tds.Ecto.ensure_all_started(TestRepo, :temporary)
{:ok, _} = Tds.Ecto.ensure_all_started(PoolRepo, :temporary)

_ = Tds.Ecto.storage_down(TestRepo.config())
:ok = Tds.Ecto.storage_up(TestRepo.config())

{:ok, _pid} = TestRepo.start_link()
{:ok, _pid} = PoolRepo.start_link()

:ok = Ecto.Migrator.up(TestRepo, 0, Ecto.Integration.Migration, log: false)
Ecto.Adapters.SQL.Sandbox.mode(TestRepo, :manual)

# :dbg.stop_clear()

Process.flag(:trap_exit, true)


mix.exs
defmodule TDS.Ecto.Mixfile do
  use Mix.Project

  @version "2.2.1"

  def project() do
    [
      app: :tds_ecto,
      version: @version,
      elixir: "~> 1.0",
      deps: deps(),
      description: description(),
      package: package()
   ]
  end

  # Configuration for the OTP application
  #
  # Type `mix help compile.app` for more information
  def application() do
    [applications: [:tds, :ecto]]
  end

  defp deps() do
    [
      {:ecto, ">= 2.0.0"},
      {:tds, "~> 1.1"},
      # {:tds, path: "../tds"},
      {:ex_doc, ">= 0.0.0", only: :dev},
      {:poison, ">= 0.0.0", only: :test}
    ]
  end

  defp description() do
    """
    Ecto 2 Adapter for Microsoft SQL Server
    """
  end

  defp package() do
    [ name: "tds_ecto",
      files: ["lib", "mix.exs", "README*"],
      maintainers: ["Justin Schneck", "Eric Witchin", "Milan Jaric"],
      licenses: ["Apache 2.0"],
      links: %{"Github" => "https://github.com/livehelpnow/tds_ecto"}]
  end
end


lib/tds/uuid.ex
defmodule Tds.UUID do
  @moduledoc """
  An Ecto type for UUIDs strings.
  """

  @behaviour Ecto.Type

  @doc """
  The Ecto type.
  """
  def type(), do: :uuid

  @doc """
  Casts to UUID.
  """
  def cast(<< a1, a2, a3, a4, a5, a6, a7, a8, ?-,
              b1, b2, b3, b4, ?-,
              c1, c2, c3, c4, ?-,
              d1, d2, d3, d4, ?-,
              e1, e2, e3, e4, e5, e6, e7, e8, e9, e10, e11, e12 >>) do
    << c(a1), c(a2), c(a3), c(a4), c(a5), c(a6), c(a7), c(a8), ?-,
       c(b1), c(b2), c(b3), c(b4), ?-,
       c(c1), c(c2), c(c3), c(c4), ?-,
       c(d1), c(d2), c(d3), c(d4), ?-,
       c(e1), c(e2), c(e3), c(e4), c(e5), c(e6), c(e7), c(e8), c(e9), c(e10), c(e11), c(e12) >>
  catch
    :error -> :error
  else
    casted -> {:ok, casted}
  end
  def cast(<< _::128 >> = binary), do: encode(binary)
  def cast(_), do: :error

  @doc """
  Same as `cast/1` but raises `Ecto.CastError` on invalid arguments.
  """
  def cast!(value) do
    case cast(value) do
      {:ok, uuid} -> uuid
      :error -> raise Ecto.CastError, type: __MODULE__, value: value
    end
  end

  @compile {:inline, c: 1}

  defp c(?0), do: ?0
  defp c(?1), do: ?1
  defp c(?2), do: ?2
  defp c(?3), do: ?3
  defp c(?4), do: ?4
  defp c(?5), do: ?5
  defp c(?6), do: ?6
  defp c(?7), do: ?7
  defp c(?8), do: ?8
  defp c(?9), do: ?9
  defp c(?A), do: ?a
  defp c(?B), do: ?b
  defp c(?C), do: ?c
  defp c(?D), do: ?d
  defp c(?E), do: ?e
  defp c(?F), do: ?f
  defp c(?a), do: ?a
  defp c(?b), do: ?b
  defp c(?c), do: ?c
  defp c(?d), do: ?d
  defp c(?e), do: ?e
  defp c(?f), do: ?f
  defp c(_),  do: throw(:error)

  @doc """
  Converts a string representing a UUID into a binary.
  """
  def dump(<< a1, a2, a3, a4, a5, a6, a7, a8, ?-,
              b1, b2, b3, b4, ?-,
              c1, c2, c3, c4, ?-,
              d1, d2, d3, d4, ?-,
              e1, e2, e3, e4, e5, e6, e7, e8, e9, e10, e11, e12 >>) do
    try do
      << d(a7)::4, d(a8)::4, d(a5)::4, d(a6)::4,
         d(a3)::4, d(a4)::4, d(a1)::4, d(a2)::4,
         d(b3)::4, d(b4)::4, d(b1)::4, d(b2)::4,
         d(c3)::4, d(c4)::4, d(c1)::4, d(c2)::4,
         d(d1)::4, d(d2)::4, d(d3)::4, d(d4)::4,
         d(e1)::4, d(e2)::4, d(e3)::4, d(e4)::4,
         d(e5)::4, d(e6)::4, d(e7)::4, d(e8)::4,
         d(e9)::4, d(e10)::4, d(e11)::4, d(e12)::4 >>
    catch
      :error -> :error
    else
      binary ->
        {:ok, binary}
    end
  end
  def dump(_), do: :error

  @compile {:inline, d: 1}

  defp d(?0), do: 0
  defp d(?1), do: 1
  defp d(?2), do: 2
  defp d(?3), do: 3
  defp d(?4), do: 4
  defp d(?5), do: 5
  defp d(?6), do: 6
  defp d(?7), do: 7
  defp d(?8), do: 8
  defp d(?9), do: 9
  defp d(?A), do: 10
  defp d(?B), do: 11
  defp d(?C), do: 12
  defp d(?D), do: 13
  defp d(?E), do: 14
  defp d(?F), do: 15
  defp d(?a), do: 10
  defp d(?b), do: 11
  defp d(?c), do: 12
  defp d(?d), do: 13
  defp d(?e), do: 14
  defp d(?f), do: 15
  defp d(_),  do: throw(:error)

  @doc """
  Converts a binary UUID into a string.
  """
  def load(<<_::128>> = uuid) do
    encode(uuid)
  end
  def load(<<_::64, ?-, _::32, ?-, _::32, ?-, _::32, ?-, _::96>> = string) do
    raise ArgumentError, "trying to load string UUID as Tds.UUID: #{inspect string}. " <>
                         "Maybe you wanted to declare :uuid as your database field?"
  end
  def load(%Ecto.Query.Tagged{type: :uuid, value: uuid}) do
    encode(uuid)
  end
  def load(_), do: :error

  @doc """
  Generates a version 4 (random) UUID.
  """
  def generate do
    {:ok, uuid} = encode(bingenerate())
    uuid
  end

  @doc """
  Generates a version 4 (random) UUID in the binary format.
  """
  def bingenerate do
    << a1::4, a2::4, a3::4, a4::4,
       a5::4, a6::4, a7::4, a8::4,
       b1::4, b2::4, b3::4, b4::4,
       _ ::4, c2::4, c3::4, c4::4,
       d1::4, d2::4, d3::4, d4::4,
       e1::4, e2::4, e3::4, e4::4,
       _ ::4, e6::4, e7::4, e8::4,
       e9::4, e10::4, e11::4, e12::4 >> = :crypto.strong_rand_bytes(16)
    << a7::4, a8::4, a5::4, a6::4,
       a3::4, a4::4, a1::4, a2::4,
       b3::4, b4::4, b1::4, b2::4,
       c3::4, c4::4, 4 ::4, c2::4,
       d1::4, d2::4, d3::4, d4::4,
       e1::4, e2::4, e3::4, e4::4,
       2 ::4, e6::4, e7::4, e8::4,
       e9::4, e10::4, e11::4, e12::4 >>
  end

  # Callback invoked by autogenerate fields.
  @doc false
  def autogenerate, do: generate()

  defp encode(<< a1::4, a2::4, a3::4, a4::4,
                 a5::4, a6::4, a7::4, a8::4,
                 b1::4, b2::4, b3::4, b4::4,
                 c1::4, c2::4, c3::4, c4::4,
                 d1::4, d2::4, d3::4, d4::4,
                 e1::4, e2::4, e3::4, e4::4,
                 e5::4, e6::4, e7::4, e8::4,
                 e9::4, e10::4, e11::4, e12::4 >>) do
    << e(a7), e(a8), e(a5), e(a6), e(a3), e(a4), e(a1), e(a2), ?-,
       e(b3), e(b4), e(b1), e(b2), ?-,
       e(c3), e(c4), e(c1), e(c2), ?-,
       e(d1), e(d2), e(d3), e(d4), ?-,
       e(e1), e(e2), e(e3), e(e4), e(e5), e(e6), e(e7), e(e8), e(e9), e(e10), e(e11), e(e12) >>
  catch
    :error -> :error
  else
    encoded -> {:ok, encoded}
  end

  @compile {:inline, e: 1}

  defp e(0),  do: ?0
  defp e(1),  do: ?1
  defp e(2),  do: ?2
  defp e(3),  do: ?3
  defp e(4),  do: ?4
  defp e(5),  do: ?5
  defp e(6),  do: ?6
  defp e(7),  do: ?7
  defp e(8),  do: ?8
  defp e(9),  do: ?9
  defp e(10), do: ?a
  defp e(11), do: ?b
  defp e(12), do: ?c
  defp e(13), do: ?d
  defp e(14), do: ?e
  defp e(15), do: ?f
end


lib/tds_ecto/migration.ex
defmodule Tds.Ecto.Migration do

  alias Ecto.Migration.{Table, Index, Reference, Constraint}

  def execute_ddl({command, %Table{} = table, columns}) when command in [:create, :create_if_not_exists] do
    table_structure =
      case column_definitions(table, columns) ++ pk_definitions(columns, ", CONSTRAINT [PK_#{table.prefix}_#{table.name}] ") do
        [] -> []
        list -> [?\s, ?(, list, ?)]
      end

    [[if_table_not_exists(command == :create_if_not_exists, table.name, table.prefix),
      "CREATE TABLE ",
      quote_table(table.prefix, table.name),
      table_structure,
      engine_expr(table.engine),
      options_expr(table.options),
      if_do(command == :create_if_not_exists, "END ")]]
  end

  def execute_ddl({command, %Table{} = table}) when command in [:drop, :drop_if_exists] do
    [[if_table_exists(command == :drop_if_exists, table.name, table.prefix),
      "DROP TABLE ",
      quote_table(table.prefix, table.name),
      if_do(command == :drop_if_exists, "END ")]]
  end

  def execute_ddl({:alter, %Table{} = table, changes}) do
    [["ALTER TABLE ", quote_table(table.prefix, table.name), ?\s,
      column_changes(table, changes),
      pk_definitions(changes, ", ADD CONSTRAINT [PK_#{table.prefix}_#{table.name}] ")]]
  end

  def execute_ddl({:create, %Index{} = index}) do
    if index.where do
      error!(nil, "MySQL adapter does not support where in indexes")
    end

    [["CREATE", if_do(index.unique, " UNIQUE"), " INDEX ",
      quote_name(index.name),
      " ON ",
      quote_table(index.prefix, index.table), ?\s,
      ?(, intersperse_map(index.columns, ", ", &index_expr/1), ?),
      if_do(index.using, [" USING ", to_string(index.using)]),
      if_do(index.concurrently, " LOCK=NONE")]]
  end

  def execute_ddl({:create_if_not_exists, %Index{}}),
    do: error!(nil, "MySQL adapter does not support create if not exists for index")

  def execute_ddl({:create, %Constraint{check: check}}) when is_binary(check),
    do: error!(nil, "MySQL adapter does not support check constraints")
  def execute_ddl({:create, %Constraint{exclude: exclude}}) when is_binary(exclude),
    do: error!(nil, "MySQL adapter does not support exclusion constraints")

  def execute_ddl({:drop, %Index{} = index}) do
    [["DROP INDEX ",
      quote_name(index.name),
      " ON ", quote_table(index.prefix, index.table),
      if_do(index.concurrently, " LOCK=NONE")]]
  end

  def execute_ddl({:drop, %Constraint{}}),
    do: error!(nil, "MySQL adapter does not support constraints")

  def execute_ddl({:drop_if_exists, %Index{}}),
    do: error!(nil, "MySQL adapter does not support drop if exists for index")

  def execute_ddl({:rename, %Table{} = current_table, %Table{} = new_table}) do
    [["RENAME TABLE ", quote_table(current_table.prefix, current_table.name),
      " TO ", quote_table(new_table.prefix, new_table.name)]]
  end

  def execute_ddl({:rename, _table, _current_column, _new_column}) do
    error!(nil, "MySQL adapter does not support renaming columns")
  end

  def execute_ddl(string) when is_binary(string), do: [string]

  def execute_ddl(keyword) when is_list(keyword),
    do: error!(nil, "MySQL adapter does not support keyword lists in execute")

  defp pk_definitions(columns, prefix) do
    pks =
      for {_, name, _, opts} <- columns,
          opts[:primary_key],
          do: name

    case pks do
      [] -> []
      _  -> [[prefix, "PRIMARY KEY CLUSTERED (", intersperse_map(pks, ", ", &quote_name/1), ?)]]
    end
  end

  defp column_definitions(table, columns) do
    intersperse_map(columns, ", ", &column_definition(table, &1))
  end

  defp column_definition(table, {:add, name, %Reference{} = ref, opts}) do
    [quote_name(name), ?\s, reference_column_type(ref.type, opts),
     column_options(opts), reference_expr(ref, table, name)]
  end

  defp column_definition(_table, {:add, name, type, opts}) do
    [quote_name(name), ?\s, column_type(type, opts), column_options(opts)]
  end

  defp column_changes(table, columns) do
    intersperse_map(columns, ", ", &column_change(table, &1))
  end

  defp column_change(table, {:add, name, %Reference{} = ref, opts}) do
    ["ADD ", quote_name(name), ?\s, reference_column_type(ref.type, opts),
     column_options(opts), constraint_expr(ref, table, name)]
  end

  defp column_change(_table, {:add, name, type, opts}) do
    ["ADD ", quote_name(name), ?\s, column_type(type, opts), column_options(opts)]
  end

  defp column_change(table, {:modify, name, %Reference{} = ref, opts}) do
    ["MODIFY ", quote_name(name), ?\s, reference_column_type(ref.type, opts),
     column_options(opts), constraint_expr(ref, table, name)]
  end

  defp column_change(_table, {:modify, name, type, opts}) do
    ["MODIFY ", quote_name(name), ?\s, column_type(type, opts), column_options(opts)]
  end

  defp column_change(_table, {:remove, name}), do: ["DROP ", quote_name(name)]

  defp column_options(opts) do
    default = Keyword.fetch(opts, :default)
    null    = Keyword.get(opts, :null)
    [default_expr(default), null_expr(null)]
  end

  defp null_expr(false), do: " NOT NULL"
  defp null_expr(true), do: " NULL"
  defp null_expr(_), do: []

  defp default_expr({:ok, nil}),
    do: " DEFAULT NULL"
  defp default_expr({:ok, literal}) when is_binary(literal),
    do: [" DEFAULT '", escape_string(literal), ?']
  defp default_expr({:ok, literal}) when is_number(literal) or is_boolean(literal),
    do: [" DEFAULT ", to_string(literal)]
  defp default_expr({:ok, {:fragment, expr}}),
    do: [" DEFAULT ", expr]
  defp default_expr(:error),
    do: []

  defp index_expr(literal) when is_binary(literal),
    do: literal
  defp index_expr(literal), do: quote_name(literal)

  defp engine_expr(storage_engine),
    do: [" ENGINE = ", String.upcase(to_string(storage_engine || "INNODB"))]

  defp options_expr(nil),
    do: []
  defp options_expr(keyword) when is_list(keyword),
    do: error!(nil, "MySQL adapter does not support keyword lists in :options")
  defp options_expr(options),
    do: [?\s, to_string(options)]

  defp column_type(type, opts) do
    size      = Keyword.get(opts, :size)
    precision = Keyword.get(opts, :precision)
    scale     = Keyword.get(opts, :scale)
    type_name = ecto_to_db(type)

    cond do
      size            -> [type_name, ?(, to_string(size), ?)]
      precision       -> [type_name, ?(, to_string(precision), ?,, to_string(scale || 0), ?)]
      type == :string -> [type_name, "(255)"]
      true            -> type_name
    end
  end

  defp constraint_expr(%Reference{} = ref, table, name),
    do: [", ADD CONSTRAINT ", reference_name(ref, table, name),
         " FOREIGN KEY (", quote_name(name), ?),
         " REFERENCES ", quote_table(table.prefix, ref.table),
         ?(, quote_name(ref.column), ?),
         reference_on_delete(ref.on_delete), reference_on_update(ref.on_update)]

  defp reference_expr(%Reference{} = ref, table, name),
    do: [", CONSTRAINT ", reference_name(ref, table, name),
         " FOREIGN KEY (", quote_name(name), ?),
         " REFERENCES ", quote_table(table.prefix, ref.table),
         ?(, quote_name(ref.column), ?),
         reference_on_delete(ref.on_delete), reference_on_update(ref.on_update)]

  defp reference_name(%Reference{name: nil}, table, column),
    do: quote_name("#{table.name}_#{column}_fkey")
  defp reference_name(%Reference{name: name}, _table, _column),
    do: quote_name(name)

  defp reference_column_type(:serial, _opts), do: "BIGINT UNSIGNED"
  defp reference_column_type(:bigserial, _opts), do: "BIGINT UNSIGNED"
  defp reference_column_type(type, opts), do: column_type(type, opts)

  defp reference_on_delete(:nilify_all), do: " ON DELETE SET NULL"
  defp reference_on_delete(:delete_all), do: " ON DELETE CASCADE"
  defp reference_on_delete(_), do: []

  defp reference_on_update(:nilify_all), do: " ON UPDATE SET NULL"
  defp reference_on_update(:update_all), do: " ON UPDATE CASCADE"
  defp reference_on_update(_), do: []

  ## Helpers

  defp get_source(query, sources, ix, source) do
    {expr, name, _schema} = elem(sources, ix)
    {expr || paren_expr(source, sources, query), name}
  end

  defp quote_name(name)
  defp quote_name(name) when is_atom(name),
    do: quote_name(Atom.to_string(name))
  defp quote_name(name) do
    if String.contains?(name, "[") or String.contains?(name, "]") do
      error!(nil, "bad field name #{inspect name} '[' and ']' are not permited")
    end

    [?[, name, ?]]
  end

  defp quote_table(nil, name),    do: quote_table(name)
  defp quote_table(prefix, name), do: [quote_table(prefix), ?., quote_table(name)]

  defp quote_table(name) when is_atom(name),
    do: quote_table(Atom.to_string(name))
  defp quote_table(name) do
    if String.contains?(name, "[") or String.contains?(name, "]") do
      error!(nil, "bad table name #{inspect name} '[' and ']' are not permited")
    end
    [?[, name, ?]]
  end

  defp intersperse_map(list, separator, mapper, acc \\ [])
  defp intersperse_map([], _separator, _mapper, acc),
    do: acc
  defp intersperse_map([elem], _separator, mapper, acc),
    do: [acc | mapper.(elem)]
  defp intersperse_map([elem | rest], separator, mapper, acc),
    do: intersperse_map(rest, separator, mapper, [acc, mapper.(elem), separator])

  defp if_do(condition, value) do
    if condition, do: value, else: []
  end

  defp escape_string(value) when is_binary(value) do
    value
    |> :binary.replace("'", "''", [:global])
  end

  defp ecto_cast_to_db(type, query), do: ecto_to_db(type, query)

  defp ecto_to_db(type, query \\ nil)
  defp ecto_to_db({:array, _}, query),
    do: error!(query, "Array type is not supported by TDS")
  defp ecto_to_db(:id, _query),             do: "bigint"
  defp ecto_to_db(:serial, _query),         do: "int"
  defp ecto_to_db(:bigserial, _query),      do: "bigint"
  defp ecto_to_db(:binary_id, _query),      do: "uniqueidentifier"
  defp ecto_to_db(:boolean, _query),        do: "bit"
  defp ecto_to_db(:string, _query),         do: "nvarchar"
  defp ecto_to_db(:float, _query),          do: "float"
  defp ecto_to_db(:binary, _query),         do: "varbinary"
  defp ecto_to_db(:uuid, _query),           do: "uniqueidentifier"
  defp ecto_to_db({:map, :string}, _query), do: "nvarchar"
  defp ecto_to_db(:utc_datetime, _query),   do: "datetime2"
  defp ecto_to_db(:naive_datetime, _query), do: "datetime"
  defp ecto_to_db(other, _query),           do: Atom.to_string(other)

  defp error!(nil, message) do
    raise ArgumentError, message
  end
  defp error!(query, message) do
    raise Ecto.QueryError, query: query, message: message
  end

  defp if_table_not_exists(condition, name, prefix \\ "dbo") do
    if condition do
      query_segment = ["IF NOT EXISTS ( ",
                       "SELECT * ",
                       "FROM [INFORMATION_SCHEMA].[TABLES] info ",
                       "WHERE info.[TABLE_NAME] = '#{name}' ",
                       "AND info.[TABLE_SCHEMA] = '#{prefix}' ",
                       ") BEGIN "]
      Enum.map_join(query_segment, "", &"#{&1}")
    else
      []
    end
  end

  defp if_table_exists(condition, name, prefix \\ "dbo") do
    if condition do
      query_segment = ["IF EXISTS ( ",
                       "SELECT * ",
                       "FROM [INFORMATION_SCHEMA].[TABLES] info ",
                       "WHERE info.[TABLE_NAME] = '#{name}' ",
                       "AND info.[TABLE_SCHEMA] = '#{prefix}' ",
                       ") BEGIN "]
      Enum.map_join(query_segment, "", &"#{&1}")
    else
      []
    end
  end

  def uuid(<<v1::32, v2::16, v3::16, v4::64>>) do
    <<v1::little-signed-32, v2::little-signed-16, v3::little-signed-16, v4::signed-64>>
  end

  

end